presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchManager.java

```diff
@@ -268,16 +270,18 @@ private <C> void createQueryInternal(QueryId queryId, String slug, int retryCoun
             }
 
             // decode session
-            session = sessionSupplier.createSession(queryId, sessionContext, warningCollectorFactory);
+            sessionBuilder = sessionSupplier.createSessionBuilder(queryId, sessionContext, warningCollectorFactory);
+            session = sessionBuilder.build();
 
             // prepare query
-            AnalyzerOptions analyzerOptions = createAnalyzerOptions(session, session.getWarningCollector());
+            AnalyzerOptions analyzerOptions = createAnalyzerOptions(session, sessionBuilder.getWarningCollector());
             QueryPreparerProvider queryPreparerProvider = queryPreparerProviderManager.getQueryPreparerProvider(getAnalyzerType(session));
-            preparedQuery = queryPreparerProvider.getQueryPreparer().prepareQuery(analyzerOptions, query, session.getPreparedStatements(), session.getWarningCollector());
+            preparedQuery = queryPreparerProvider.getQueryPreparer().prepareQuery(analyzerOptions, query, sessionBuilder.getPreparedStatements(), sessionBuilder.getWarningCollector());
             query = preparedQuery.getFormattedQuery().orElse(query);
 
             // select resource group
             Optional<QueryType> queryType = preparedQuery.getQueryType();
+            sessionBuilder.setQueryType(queryType);
```

Comment:
why make all these changes instead when you can just do 
`session = Session.builder(session).setQueryType(queryType).build();`

Reply:
This is because of this [line](https://github.com/prestodb/presto/blob/2f4ea64d3bf3fa8c89b3e805d5d0987ac6d971c2/presto-main/src/main/java/com/facebook/presto/Session.java#L644). My change makes it so we still have access to the SessionBuilder object before we begin the transaction. After we start the transaction, we can't use the ctor that passes in Session. I could use the other [ctor](https://github.com/prestodb/presto/blob/2f4ea64d3bf3fa8c89b3e805d5d0987ac6d971c2/presto-main/src/main/java/com/facebook/presto/Session.java#L598) but would need to refill in all the Session fields.